doc/spec/Types.md

```diff
@@ -1322,6 +1352,81 @@ foo({ id: 1234, name: false });    // Error, name of wrong type
 foo({ name: "hello" });            // Error, id required but missing
 ```
 
+### Comparability
+
+*TODO: Document the base primitive type.*
+
+Types are required to be *comparable* in certain circumstances, such as part of when checking whether two values of given types might be equal at runtime using operators like '===', or when using a type assertion.
+
+*S* is ***comparable to*** a type *T*, and *T* is ***comparable from*** *S*, if *S* has no excess properties with respect to *T* ([#excess-properties]<!--3.11.5-->) and one of the following is true:
+
+* *S* and *T* are identical types.
+* *S* or *T* is the Any type.
+* *S* is the Never type.
+* *S* is the Undefined type and *T* is the Void type.
+* *T* is the Undefined type and *S* is the Void type.
+* *S* is either the Undefined or Null type, and strict null checks ([#strict-null-checks]) are disabled.
+* *S* is an object type and *T* is the non-primitive `object` type.
+* *T* is an object type and *S* is the non-primitive `object` type.
+* *S* or *T* is a numeric enum or numeric enum member type and the other is the primitive type Number.
+* *S* and *T* are both enum member types with respective containing enum types *E* and *F*, and
+    * *E* is comparable to *F*.
+    * *F* is comparable to *E*.
+* *S* and *T* are both non-const enum types with the same declared name, and
+  * for each enum member *M* in *E*, there exists an enum member declaration *N* in *F* such that *M* and *N* have the same declared name and value.
+  * for each enum member *N* in *F*, there exists an enum member declaration *M* in *E* such that *N* and *M* have the same declared name and value.
+* *S* is an enum member type, *T* is a literal type, and the literal values of *S* and *T* are identical.
+* *T* is an enum member type, *S* is a literal type, and the literal values of *S* and *T* are identical.
+* *S* is a literal type and *T* is the base primitive type of *S*.
+* *T* is a literal type and *S* is the base primitive type of *T*.
+* *S* is a union type and some constituent type of *S* is comparable to *T*.
+* *S* is an intersection type and at least one constituent type of *S* is comparable to *T*.
+* *T* is a union type and *S* is comparable to at least one constituent type of *T*.
+* *T* is an intersection type and *S* is comparable to each constituent type of *T*.
+* *T* is a type parameter and *S* is a mapped type *{ [P in keyof X]: Y }*, and *Y* is comparable to *X[P]*.

```

Comment:
I think that you have to say that `X` is comparable (or identical?) to `T` if you don't just use `keyof T` in the mapped type itself.